data-products/src/main/scala/org/sunbird/analytics/util/TextBookUtils.scala

```diff
@@ -0,0 +1,183 @@
+    })
+
+    val dceRDD=dceTextBookRDD.map(e=>(e.identifier,e))
+    val etbRDD=etbTextBookRDD.map(e=>(e.identifier,e)).fullOuterJoin(dceRDD)
+    val finalOutput = etbRDD.map(e=> FinalOutput(e._1,e._2._1,e._2._2))
+
+    finalOutput
+  }
+
+  def generateDCETextbookReport(response: ContentInfo)(implicit sc: SparkContext): List[dceTextbookData] = {
+    var index=0
+    var totalQRCodes=0
+    var qrLinked=0
+    var qrNotLinked=0
+    var term1NotLinked=0
+    var term2NotLinked=0
+    var dceReport = List[dceTextbookData]()
+
+    if(response!=null && response.children.size>0 && response.status=="Live") {
+      val lengthOfChapters = response.children.get.length
+      val term = if(index<=lengthOfChapters/2) "T1"  else "T2"
+      index = index+1
+
+      val dceTextbook = parseDCETextbook(response.children.get,term,0,0,0,0,0)
+      totalQRCodes = dceTextbook._2
+      qrLinked = dceTextbook._3
+      qrNotLinked = dceTextbook._4
+      term1NotLinked = dceTextbook._5
+      term2NotLinked = dceTextbook._6
+      val medium = if(response.medium!=null) response.medium.asInstanceOf[List[String]].mkString(",") else ""
+      val subject = if(response.subject!=null) response.subject.asInstanceOf[List[String]].mkString(",") else ""
+      val gradeLevel = if(response.gradeLevel!=null) response.gradeLevel.mkString(",") else ""
+      val dceDf = dceTextbookData(response.channel,response.identifier, response.name, medium, gradeLevel, subject,response.createdOn.substring(0,10), response.lastUpdatedOn.substring(0,10),totalQRCodes,qrLinked,qrNotLinked,term1NotLinked,term2NotLinked)
+      dceReport = dceDf::dceReport
+    }
+    dceReport
+  }
+
+  def parseDCETextbook(data: List[ContentInfo], term: String, counter: Integer,linkedQr: Integer, qrNotLinked:Integer, counterT1:Integer, counterT2:Integer): (Integer,Integer,Integer,Integer,Integer,Integer) = {
+    var counterValue=counter
+    var counterQrLinked = linkedQr
+    var counterNotLinked = qrNotLinked
+    var term1NotLinked = counterT1
+    var term2NotLinked = counterT2
+    var tempValue = 0
+
+    data.map(units => {
+      if(units.dialcodes!=null){
+        counterValue=counterValue+1
+
+        if(units.leafNodesCount>0) { counterQrLinked=counterQrLinked+1 }
+        else {
+          counterNotLinked=counterNotLinked+1
+          if(term == "T1") { term1NotLinked=term1NotLinked+1 }
+          else { term2NotLinked = term2NotLinked+1 }
+        }
+      }
+      if(units.contentType.get== "TextBookUnit"){
+        val output = parseDCETextbook(units.children.getOrElse(List[ContentInfo]()),term,counterValue,counterQrLinked,counterNotLinked,term1NotLinked,term2NotLinked)
+        tempValue = output._1
+        counterQrLinked = output._3
+        counterNotLinked = output._4
+        term1NotLinked = output._5
+        term2NotLinked = output._6
+      }
+    })
+    (counterValue,tempValue,counterQrLinked,counterNotLinked,term1NotLinked,term2NotLinked)
+  }
```

Comment:
I have doubt on the entire code. During recursive functionality shouldn't we add the values? instead of replacing values